docs/zpure/index.md

```diff
@@ -0,0 +1,350 @@
+  interestComputation.run
+```
+
+## Working With State
+
+The state type described by `S1` and `S2` is the first capability provided by `ZPure` that is not directly analogous to `ZIO`.
+
+In `ZIO` we manage state using the `Ref` and `FiberRef` data types so we do not need these additional type parameters. In contrast, in `ZPure` we want to avoid arbitrary side effects so we instead model state as a function that takes the old state and returns the new state using the `S1` and `S2` type parameters.
+
+In practice in most cases `S1` and `S2` will be the same and we will be describing a computation that takes some state and just returns an updated value of the same state type. However, in some cases it can be useful to model the initial and updated state separately, for example to describe a transition from a `Closed` to `Open` state in a type safe way.
+
+To motivate our example here let's define an `AccountState` type.
+
+```scala mdoc
+case class AccountState(balance: Int, open: Boolean)
+```
+
+We will also define a domain specific error type which we will use.
+
+```scala mdoc
+sealed trait AccountError
+
+case object InsufficientFunds extends AccountError
+```
+
+The most basic operators for working with state are `get` and `set`. The `get` operator retrieves the initial state and the `set` operator sets the updated state to the specified value.
+
+Let's use these operators to describe a computation that withdraws funds from the account.
+
+```scala mdoc
+def withdraw(amount: Int): ZPure[Nothing, AccountState, AccountState, Any, AccountError, Unit] =
+  for {
+    state <- ZPure.get[AccountState]
+    _     <- if (amount > state.balance) ZPure.fail(InsufficientFunds)
+             else ZPure.set(AccountState(state.balance - amount, state.open))
+  } yield ()
+```
+
+This example combines a couple of the features we have talked about so far.
+
+First, we are using a _for comprehension_ to chain together multiple computations
+
+Second, we are using both the error type and the state type here. If the customer has insufficient funds we fail immediately, otherwise we update the state.
+
+One important thing to note here is that since `ZPure` computations are never concurrent we don't need to worry about conflicting modifications to the state.
+
+For example, in the context of `ZIO` code analogous to the above would not be safe because the `get` and `set` would not be performed atomically so we would need to do the entire transaction in a single `modify` operation, returning a value indicating whether there were sufficient funds. In `ZPure` we don't have to do any of that so our code can be simpler.
+
+There are also `update` and `modify` operators on `ZPure` so for example we could factor out the logic of updating the account balance like this:
+
+```scala mdoc
+def decrementBalance(amount: Int): ZPure[Nothing, AccountState, AccountState, Any, Nothing, Unit] =
+  ZPure.update(state => AccountState(state.balance - amount, state.open))
+```
```

Comment:
I presume that you intend for `decrementBalance(amount: Int)` to replace the `else` clause in the `withdraw` example above?  If so, seems like it it would be nice to see that here prior to the run.

Reply:
Done!